src/libraries/Microsoft.PowerFx.Json/Functions/JsonFunctionImpl.cs

```diff
@@ -129,17 +143,58 @@ private JsonFlags GetFlags()
 
             private class Utf8JsonWriterVisitor : IValueVisitor
             {
+                private const int _maxDepth = 20;           // maximum depth of UO
+
+                private const int _maxLength = 1024 * 1024; // 1 MB, maximum number of bytes allowed to be sent to Utf8JsonWriter
+
                 private readonly Utf8JsonWriter _writer;
+
                 private readonly TimeZoneInfo _timeZoneInfo;
+
                 private readonly bool _flattenValueTables;
 
+                private readonly Canceller _canceller;
+
                 internal readonly List<ErrorValue> ErrorValues = new List<ErrorValue>();
 
-                internal Utf8JsonWriterVisitor(Utf8JsonWriter writer, TimeZoneInfo timeZoneInfo, bool flattenValueTables)
+                internal Utf8JsonWriterVisitor(Utf8JsonWriter writer, TimeZoneInfo timeZoneInfo, bool flattenValueTables, Canceller canceller)
                 {
                     _writer = writer;
                     _timeZoneInfo = timeZoneInfo;
                     _flattenValueTables = flattenValueTables;
+
+                    _canceller = canceller;
+                }
+
+                private void ThrowIfCancellationRequested(int index)
```

Comment:
>ThrowIfCancellationRequested [](http://example.com/codeflow?start=29&length=28)

Suggestion: change the method name to something like "CheckLimitsAndCancellation" as this doesn't only check for cancellation